### DISCUSSION TITLE
Can you use an extension library only at a subdirectory?

### DISCUSSION THREAD

**Discussion by stdedos** [rubocop/rubocop#13306] [Q&A]
I want to use rubocop-rails, but only in a subdirectory.

My layout is

```console
$ tree . --prune -aP '*rubocop*' -I build
.
├── backend
│   ├── .rubocop.yml
│   └── vendor
│       └── cache
│           ├── rubocop-1.66.1.gem
│           ├── rubocop-ast-1.32.3.gem
│           ├── rubocop-performance-1.22.1.gem
│           └── rubocop-rails-2.26.2.gem
├── ci-tools
│   └── .rubocop.yml
└── .rubocop.yml
```

and _for reasons_ I **must** call rubocop as `GIT_TOP_LEVEL/backend$ bin/bundle exec rubocop -a ..`
(that seem unable to get around)

-----------

I tried to 

```yml
diff --git a/.rubocop.yml b/.rubocop.yml
index f64c6ae1a92..b101e2ef15a 100644
--- a/.rubocop.yml
+++ b/.rubocop.yml
@@ -1,6 +1,5 @@
 require:
   - rubocop-performance
-  - rubocop-rails
 
 AllCops:
   ActiveSupportExtensionsEnabled: true
diff --git a/backend/.rubocop.yml b/backend/.rubocop.yml
index 939c4129dae..91e112f1e51 100644
--- a/backend/.rubocop.yml
+++ b/backend/.rubocop.yml
@@ -1,3 +1,6 @@
+require:
+  - rubocop-rails
+
 # We still want to apply the global RuboCop settings in the project root when
 # running RuboCop from within the backend directory (e.g. by IDEA).
 inherit_from: ../.rubocop.yml
```

but it seems that one of the directories (sibling to `backend/`, not visible above) is simply "not checked".

If I do `bin/bundle exec rubocop -a ../kernel/`, then I see multiple `Rails/` rules applied.

-------------

The alternative

```yaml
Rails:
  Enabled: true
  Exclude:
    - 'kernel/**/*'
```

_which I don't know if it is valid_, fails to "scope" `Rails` cops applicability.

**Comment by Earlopain** [comment]
Can you make a sample repo that illustrates what you are trying to do? I don't understand what exactly the problem is.

**Reply by stdedos** [reply]
That might be super hard :sweat: Any of the "special-purpose bits" of the repo (including e.g. using JRuby, JRuby being "hosted" in a different directory than the top level, Gemfile not being at the repo top-level so any `bundle` command MUST be ran from that directory, etc) might be "at fault" here. I am trying to "eliminate out" rubocop being the curlprit.

e.g. rubocop-rails has the https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsindexby, which talks about `#index_by()`.
However, no other sub-directory except `backend/` "have Rails active", or "processed within a Rails scope" (idk how this works :sweat:)

So while https://docs.rubocop.org/rubocop-rails/cops_rails.html#railsindexby is a valid change for _any Rails code_, for the rest of the code (e.g. on `kernel/`) it is **not valid**, as there is no "Rails context" there.

Think of :point_up: e.g. as an activesupport extension that you **MUST** `require` at top level to use it (e.g. `require 'active_support/core_ext/object/blank'`) - otherwise "it doesn't exist".
And then you have two files - one with `require 'active_support/core_ext/object/blank'`, and one without. A (fictional?) Cop that "converts" `X` to `.blank?` is applicable ONLY to the first file, AND NOT at the second.

Or ONLY at one directory-or-below, and AND NOT at other places. That is what I am trying to solve, but with all Rails cops. Either by config-inheritance (cleaner), or with a directive (hopefully with `IncludeDir:` and not with `Exclude:`, as I can easier say "what to include" than what to exclude)

**Comment by stdedos** [comment]
Okay, I may have a more straightforward issue description:

```console
$ bat -pp .rubocop.yml 
inherit_from:
  - .rubocop_style.yml
  - .rubocop_todo.yml
inherit_mode:
  merge:
    - Exclude

require:
  - rubocop-performance

AllCops:
...
$  bat -pp backend/.rubocop.yml 
inherit_from: ../.rubocop.yml

require:
  - rubocop-rails
$ /usr/bin/time backend/bin/jruby -S rubocop -a --format fuubar
...
An error occurred while Rails/ActionFilter cop was inspecting k/....rb:360:22.
...
```

I have only `require:`d `rubocop-rails` for the `backend/` directory.
Hence, I don't understand why any `Rails/` cop would touch files in the `k/` directory.

-----

Does that make more sense @Earlopain?
